src/sentry/rules/processing/delayed_processing.py

```diff
@@ -1,36 +1,273 @@
+        cond for cond in conditions if is_condition_slow(cond)
+    ]
+
+    return slow_conditions
+
+
+def get_rules_to_groups(
+    rulegroup_to_events: list[dict[str, set[str]]]
+) -> DefaultDict[int, set[int]]:
+    rules_to_groups: DefaultDict[int, set[int]] = defaultdict(set)
+    for rulegroup_to_event in rulegroup_to_events:
+        for rule_group in rulegroup_to_event.keys():
+            rule_id, group_id = rule_group.split(":")
+            rules_to_groups[int(rule_id)].add(int(group_id))
+
+    return rules_to_groups
+
+
+def get_rule_to_slow_conditions(alert_rules: list[Rule]) -> DefaultDict[Rule, list[dict[str, Any]]]:
+    rule_to_slow_conditions: DefaultDict[Rule, list[dict[str, Any]]] = defaultdict(list)
+    for rule in alert_rules:
+        slow_conditions = get_slow_conditions(rule)
+        for condition_data in slow_conditions:
+            rule_to_slow_conditions[rule].append(condition_data)
+
+    return rule_to_slow_conditions
+
+
+def get_condition_groups(
+    alert_rules: list[Rule], rules_to_groups: DefaultDict[int, set[int]]
+) -> dict[UniqueCondition, DataAndGroups]:
+    """Map unique conditions to the group IDs that need to checked for that
+    condition. We also store a pointer to that condition's JSON so we can
+    instantiate the class later
+    """
+    condition_groups: dict[UniqueCondition, DataAndGroups] = {}
+    for rule in alert_rules:
+        # We only want a rule's slow conditions because alert_rules are only added
+        # to the buffer if we've already checked their fast conditions.
+        slow_conditions = get_slow_conditions(rule)
+        for condition_data in slow_conditions:
+            if condition_data:
+                unique_condition = UniqueCondition(
+                    condition_data["id"], condition_data["interval"], rule.environment_id
+                )
+                # Add to set of group_ids if there are already group_ids
+                # that apply to the unique condition
+                if data_and_groups := condition_groups.get(unique_condition):
+                    data_and_groups.group_ids.update(rules_to_groups[rule.id])
+                # Otherwise, create the tuple containing the condition data and the
+                # set of group_ids that apply to the unique condition
+                else:
+                    condition_groups[unique_condition] = DataAndGroups(
+                        condition_data, rules_to_groups[rule.id]
+                    )
+    return condition_groups
+
+
+def get_condition_group_results(
+    condition_groups: dict[UniqueCondition, DataAndGroups], project: Project, rule: Rule
+) -> dict[UniqueCondition, dict[int, int]]:
+    # XXX: Probably want to safe execute somewhere in this step before making
+    # the query
+    condition_group_results: dict[UniqueCondition, dict[int, int]] = {}
+    for unique_condition, (condition_data, group_ids) in condition_groups.items():
```

Comment:
A lot of this function's code is duplicated in `event_frequency.py`'s `get_rate` and we can probably refactor in a future PR to clean this up.